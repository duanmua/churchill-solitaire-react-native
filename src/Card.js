import React, { Component } from 'react';
import { View, Animated, PanResponder } from 'react-native';

class Card extends Component {
  constructor(props) {
    super(props);

    const position = new Animated.ValueXY();

    const panResponder = PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderMove: (event, gesture) => {
        position.setValue({ x: gesture.dx, y: gesture.dy });
      },
      onPanResponderRelease: () => {
          this.resetPosition();
      }
    });

    this.panResponder = panResponder;
    this.position = position;
  }

  resetPosition() {
    Animated.spring(this.position, {
      toValue: {x: 0, y: 0}
    }).start();
  }

  render() {
    return (
      <Animated.View
        {...this.panResponder.panHandlers}
        style={this.position.getLayout()}
      >
        <View style={styles.ball}></View>
      </Animated.View>
    );
  }
}

const styles = {
  ball: {
    height: 60,
    width: 60,
    //borderRadius: 30,
    borderWidth: 30,
    borderColor: 'black'
  }
}

export default Card;
