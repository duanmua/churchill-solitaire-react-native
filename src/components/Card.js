import React, { Component } from 'react';
import { View, Animated, PanResponder, Text } from 'react-native';
import { SCREEN_WIDTH, CARD_HEIGHT, CARD_OFFSET } from '../LayoutConstants';

class Card extends Component {
  constructor(props) {
    super(props);

    const position = new Animated.ValueXY();
    const { cardMoved, cardDragStart, cardId, draggable, lane, index } = this.props;

    const panResponder = PanResponder.create({
      onStartShouldSetPanResponder: (event, gesture) => {
        //console.log(gesture);
        return draggable;
      },
      onPanResponderMove: (event, gesture) => {
        position.setValue({ x: gesture.dx, y: gesture.dy });
        //console.log(gesture);
      },
      onPanResponderGrant: (event, gesture) => {
        cardDragStart({lane, index});
      },
      onPanResponderRelease: (event, gesture) => {
        this.resetPosition();
        cardMoved({
          endX: gesture.moveX,
          endY: gesture.moveY,
          cardId: cardId,
          lane: lane,
          index: index
        });
      }
    });

    this.panResponder = panResponder;
    this.position = position;
  }

  resetPosition() {
    Animated.spring(this.position, {
      toValue: {x: 0, y: 0}
    }).start();
  }

  getCardView() {
    const { shown, value, suit, color } = this.props;
    let backgroundColor = shown ? 'white':'grey';
    let display = shown ? '':'none';
    return (
      <View style={{backgroundColor,  ...styles.card}}>
        <Text
          style={ {display, color, ...styles.text} }
        >
          {value.value} {suit}
        </Text>
      </View>
    );
  }

  render() {
    return (
      <Animated.View
        {...this.panResponder.panHandlers}
        style={this.position.getLayout()}
      >
        {this.getCardView()}
        <View style={{ top: CARD_OFFSET }}>
          {this.props.children}
        </View>

      </Animated.View>
    );
  }
}

const styles = {
  card: {
    height: CARD_HEIGHT,
    borderRadius: 3,
    borderWidth: 1,
    borderColor: 'black'
  },
  text: {
    //color: this.props.color,
    fontSize: 10
  }
}

export default Card;
