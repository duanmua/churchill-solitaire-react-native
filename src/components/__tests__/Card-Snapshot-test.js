import React from 'react';
import renderer from 'react-test-renderer';

import Card from '../Card.js';

it('renders a Card using Snapshots', () => {
  const card = {
    id: "A4",
    value: "Q",
    suit: "â™¥",
    color: "red",
    shown: true,
    draggable: true
  };
  expect(renderer.create(
    <Card
      key={ "key" }
      cardMoved={ jest.fn() }
      cardId={ card.id }
      value={ card.value }
      suit={ card.suit }
      color={ card.color }
      shown={ card.shown }
      draggable={ card.draggable }
      lane={ 1 }
      index={ 1 }
    />
  )).toMatchSnapshot();
});
