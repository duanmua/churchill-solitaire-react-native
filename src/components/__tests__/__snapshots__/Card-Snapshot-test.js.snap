// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Card using Snapshots 1`] = `
<View
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "left": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "black",
        "borderRadius": 3,
        "borderWidth": 1,
        "height": 60,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "display": "",
          "fontSize": 10,
        }
      }
    >
       
      â™¥
    </Text>
  </View>
  <View
    style={
      Object {
        "top": -45,
      }
    }
  />
</View>
`;
